aidb/vector_database/marqo_vector_database.py

```diff
@@ -0,0 +1,215 @@
+
+    if not index_name:
+      raise Exception('Index name must not be none')
+
+    if similarity not in ['cosinesimil', 'l1', 'l2', 'linf']:
+      raise Exception('Similarity function must be one of l1, l2, linf and cosinesiml')
+
+    if index_name in self.index_list:
+      raise Exception(f'Index {index_name} already exists, please use another name')
+
+    self.marqo_client.create_index(
+      index_name=index_name,
+      settings_dict={
+                "index_defaults": {
+                    "ann_parameters":{
+                        "space_type": similarity
+                    }
+                }
+            }
+      )
+    self.index_list.append(index_name)
+
+  def load_index(self):
+    '''
+    Reload index from Marqo
+    '''
+    self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+
+  def delete_index(self, index_name: str):
+    '''
+    delete an index
+    '''
+    if index_name in self.index_list:
+      self.marqo_client.delete_index(index_name)
+      # self.index_list.remove(index_name)
+      self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+      logger.info("Index '%s' deleted.", index_name)
+
+  def _sanitize_index_name(self, index_name: str) -> str:
+    '''
+    index should start with a capital
+    '''
+    return index_name[0].upper() + index_name[1:]
+  
+  def _check_index_validity(self, index_name: str):
+    if index_name not in self.index_list:
+      raise Exception(f'Couldn\'t find index {index_name}, please create it first')
+    return index_name
+  
+  def generate_uuid5(self, x): 
+    return str(uuid.uuid5(uuid.NAMESPACE_DNS, str(x)))
+  
+  def insert_data(self, index_name: str, data: pd.DataFrame):
+    '''
+    insert data into an index
+    '''
+    index_name = self._check_index_validity(index_name)
+
+    data = data.to_dict(orient="records")
+    # Data will have list of dictionaries 
+    # with 'id' and 'values' as keys.
+    
+    mod_data = []
+    for d in data:
+      # print(d)
+      _id = str(d['id'])
+      _vec = d['values']
+      mod_data.append(
+        {
+          "_id": _id,
+          "aidb_data":{
+            "vector": _vec
+          }
+        }
+      )
+    response = self.marqo_client.index(
+        index_name=index_name
+        ).add_documents(
+          documents=mod_data,
+          mappings={
+            "aidb_data":{
+              "type": "custom_vector"
+            }
+          },
+          tensor_fields=["aidb_data"],
+          auto_refresh=True
+        )
+    
+  def get_embeddings_by_id(self, index_name: str, ids: np.ndarray, reload = False) -> np.ndarray:
+    '''
+    Get data by id and return results
+    '''
+    if reload:
+      self.load_index()
+    index_name = self._check_index_validity(index_name)
+    result = []
+    for id in ids:
+      fetch_response = self.marqo_client.index(
+        index_name).get_document(document_id=str(id), expose_facets=True)
```

Comment:
Can you get multiple documents simultaneously rather than one by one?

Reply:
Done